From c8b646ad9c1b813a584f7935b99d01d57d5b9e39 Mon Sep 17 00:00:00 2001
From: Eric Woudstra <ericwouds@gmail.com>
Date: Mon, 29 Jul 2024 11:49:23 +0200
Subject: [PATCH 040/116] net: sfp: Fixup for OEM SFP-2.5G-T(-R-RM) module

Change from quirk to fixup for the OEM SFP-2.5G-T module.
Add OEM SFP-2.5G-T-R-RM module.

The module is re-branded to different brands:

OEM     SFP-2.5G-T       Visible brand on label LuLeey.
OEM     SFP-2.5G-T-R-RM  No visible brand, bought from Sinovoip store
FS      SFP-2.5G-T       Fiberstore [NEEDS A DIFFERENT FIXUP]
Turris  RTSFP-2.5G       Turris [SAME FIXUP AS LULEEY]
---
 drivers/net/phy/sfp.c | 26 +++++++++-----------------
 1 file changed, 9 insertions(+), 17 deletions(-)

diff --git a/drivers/net/phy/sfp.c b/drivers/net/phy/sfp.c
index 84bef5099dda..9f3cf55653ea 100644
--- a/drivers/net/phy/sfp.c
+++ b/drivers/net/phy/sfp.c
@@ -397,6 +397,13 @@ static void sfp_fixup_rollball(struct sfp *sfp)
 	sfp->phy_t_retry = msecs_to_jiffies(1000);
 }
 
+// For 2.5GBASE-T short-reach modules
+static void sfp_fixup_oem_2_5gbaset(struct sfp *sfp)
+{
+	sfp_fixup_rollball(sfp);
+	sfp->id.base.extended_cc = SFF8024_ECC_2_5GBASE_T;
+}
+
 static void sfp_fixup_rollball_wait4s(struct sfp *sfp)
 {
 	sfp_fixup_rollball(sfp);
@@ -453,22 +460,6 @@ static void sfp_quirk_2500basex(const struct sfp_eeprom_id *id,
 	__set_bit(PHY_INTERFACE_MODE_2500BASEX, caps->interfaces);
 }
 
-static void sfp_quirk_disable_autoneg(const struct sfp_eeprom_id *id,
-				      struct sfp_module_caps *caps)
-{
-	linkmode_clear_bit(ETHTOOL_LINK_MODE_Autoneg_BIT, caps->link_modes);
-}
-
-static void sfp_quirk_oem_2_5g(const struct sfp_eeprom_id *id,
-			       struct sfp_module_caps *caps)
-{
-	/* Copper 2.5G SFP */
-	linkmode_set_bit(ETHTOOL_LINK_MODE_2500baseT_Full_BIT,
-			 caps->link_modes);
-	__set_bit(PHY_INTERFACE_MODE_2500BASEX, caps->interfaces);
-	sfp_quirk_disable_autoneg(id, caps);
-}
-
 static void sfp_quirk_ubnt_uf_instant(const struct sfp_eeprom_id *id,
 				      struct sfp_module_caps *caps)
 {
@@ -545,7 +536,8 @@ static const struct sfp_quirk sfp_quirks[] = {
 	SFP_QUIRK_F("OEM", "SFP-GE-T", sfp_fixup_ignore_tx_fault),
 
 	SFP_QUIRK_F("OEM", "SFP-10G-T", sfp_fixup_rollball_cc),
-	SFP_QUIRK_S("OEM", "SFP-2.5G-T", sfp_quirk_oem_2_5g),
+	SFP_QUIRK_F("OEM", "SFP-2.5G-T", sfp_fixup_oem_2_5gbaset),
+	SFP_QUIRK_F("OEM", "SFP-2.5G-T-R-RM", sfp_fixup_oem_2_5gbaset),
 	SFP_QUIRK_S("OEM", "SFP-2.5G-BX10-D", sfp_quirk_2500basex),
 	SFP_QUIRK_S("OEM", "SFP-2.5G-BX10-U", sfp_quirk_2500basex),
 	SFP_QUIRK_F("OEM", "RTSFP-10", sfp_fixup_rollball_cc),
-- 
2.39.5

